File: frontend/src/engine/capabilities/histogram.ts
/**
 * histogram — distribution of numeric values across bins.
 *
 * Writes a 2-col table (bin upper-bound | count) and, by default, a column
 * chart to visualize. Bins can be supplied explicitly or auto-computed via
 * Sturges' rule (ceil(log2(N) + 1)).
 */

import {
  CapabilityMeta,
  HistogramParams,
  StepResult,
  ExecutionOptions,
} from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";
import { parseNumberFlexible } from "../utils/parseNumberFlexible";

const meta: CapabilityMeta = {
  action: "histogram",
  description: "Build a histogram (FREQUENCY + bin table + optional column chart)",
  mutates: true,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.2",
};

async function handler(
  context: Excel.RequestContext,
  params: HistogramParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const {
    dataRange,
    outputRange,
    bins,
    binCount,
    includeChart = true,
    chartType = "columnClustered",
    hasHeaders = true,
  } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would build histogram for ${dataRange}` };
  }

  options.onProgress?.("Computing histogram bins...");

  // Read data to determine min/max for auto-binning.
  const dataRaw = resolveRange(context, dataRange);
  const dataUsed = dataRaw.getUsedRange(false);
  dataUsed.load(["values", "rowCount", "columnCount", "address"]);
  await context.sync();

  const values = (dataUsed.values ?? []) as (string | number | boolean | null)[][];
  const startRow = hasHeaders ? 1 : 0;
  const nums: number[] = [];
  for (let r = startRow; r < values.length; r++) {
    // Tolerate text-stored numbers ("1,234", "$100", "50%", "(100)") that
    // arrive from CSV imports — parseNumberFlexible returns null for
    // genuinely non-numeric cells and we skip those.
    const v = parseNumberFlexible(values[r][0]);
    if (v !== null) nums.push(v);
  }
  if (nums.length === 0) {
    return { stepId: "", status: "error", message: "No numeric values found in dataRange." };
  }

  // Compute bins.
  let binEdges: number[];
  if (bins && bins.length > 0) {
    binEdges = [...bins].sort((a, b) => a - b);
  } else {
    const n = nums.length;
    const count = binCount ?? Math.max(2, Math.ceil(Math.log2(n) + 1)); // Sturges
    const min = Math.min(...nums);
    const max = Math.max(...nums);
    const step = (max - min) / count;
    binEdges = [];
    for (let i = 1; i <= count; i++) binEdges.push(min + step * i);
  }

  // Prepare output block: headers + bin labels + FREQUENCY formula. FREQUENCY
  // is an array formula — the trick: enter it on a range 1 taller than the
  // bins list so the "more than last bin" count lands in the final cell.
  //
  // Layout:
  //   outputRange.topLeft   →  ["Bin", "Count"]
  //   next rows: [bin1, 0], [bin2, 0], ..., [binN, 0], [>binN, 0]
  // Then we write =FREQUENCY(dataRange, binEdgesRange) as an array into the
  // count column.
  const out = resolveRange(context, outputRange);
  out.load(["address", "worksheet/name"]);
  await context.sync();

  const outAddrPart = out.address.includes("!") ? out.address.split("!").pop()! : out.address;
  const outFirst = outAddrPart.split(":")[0].match(/^([A-Z]+)(\d+)$/);
  if (!outFirst) {
    return { stepId: "", status: "error", message: `Could not parse output address: ${out.address}` };
  }
  const outStartCol = outFirst[1];
  const outStartRow = Number(outFirst[2]);
  const outSheet = out.worksheet;

  // Build the 2D matrix: 1 header + binEdges.length + 1 overflow row.
  const rows: (string | number | null)[][] = [["Bin", "Count"]];
  for (const edge of binEdges) rows.push([edge, null]);
  rows.push([`>${binEdges[binEdges.length - 1]}`, null]);

  const outputBlock = outSheet.getRangeByIndexes(
    outStartRow - 1,
    columnLetterToIndex(outStartCol),
    rows.length,
    2,
  );
  try {
    outputBlock.values = rows as unknown as (string | number | boolean)[][];
    await context.sync();
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    return { stepId: "", status: "error", message: `Failed to write histogram block: ${msg}`, error: msg };
  }

  // Write FREQUENCY array formula into the Count column, covering all rows
  // after the header.
  const nBins = binEdges.length + 1; // include the overflow row
  const countCol = nextColumnLetter(outStartCol);
  const countFirstRow = outStartRow + 1;
  const countLastRow = outStartRow + nBins;
  const binRefFirstRow = outStartRow + 1;
  const binRefLastRow = outStartRow + binEdges.length;
  const countRangeAddr = `${countCol}${countFirstRow}:${countCol}${countLastRow}`;
  const binRangeAddr = `${outStartCol}${binRefFirstRow}:${outStartCol}${binRefLastRow}`;
  // The source data range for FREQUENCY — use the actual data region (skip header if present).
  const dataAddr = dataUsed.address;
  const dataAddrOnly = dataAddr.includes("!") ? dataAddr.split("!").pop()! : dataAddr;
  const dataSheetPrefix = dataAddr.includes("!") ? dataAddr.split("!")[0] + "!" : "";
  // Build a bounded address for just the data rows (skip header row on column 1).
  const dataMatch = dataAddrOnly.match(/^\$?([A-Z]+)\$?(\d+):\$?([A-Z]+)\$?(\d+)$/);
  let dataFormulaRef = dataAddr;
  if (dataMatch) {
    const [, c1, r1, c2, r2] = dataMatch;
    const dataStart = hasHeaders ? Number(r1) + 1 : Number(r1);
    dataFormulaRef = `${dataSheetPrefix}${c1}${dataStart}:${c2}${r2}`;
  }
  const formula = `=FREQUENCY(${dataFormulaRef},${binRangeAddr})`;

  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const countRange = outSheet.getRange(countRangeAddr) as any;
    // setFormulaArray introduced on Range; prefer formulas = [[...]] on a
    // single-column range — modern Excel treats this as an array formula
    // when the expression is FREQUENCY.
    const singleColFormula: string[][] = [];
    for (let i = 0; i < nBins; i++) singleColFormula.push([formula]);
    countRange.formulas = singleColFormula;
    await context.sync();
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    return { stepId: "", status: "error", message: `Failed to write FREQUENCY formula: ${msg}`, error: msg };
  }

  // Optional chart.
  let chartName: string | undefined;
  if (includeChart) {
    try {
      const chartDataRange = outSheet.getRange(`${outStartCol}${outStartRow}:${countCol}${countLastRow}`);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const chart: any = (outSheet as any).charts.add(chartType, chartDataRange, "auto");
      chart.title.text = "Histogram";
      chartName = chart.name ?? "Histogram";
      await context.sync();
    } catch {
      // Chart failures shouldn't abort the main output.
    }
  }

  const outputAddr = `${outSheet.name}!${outStartCol}${outStartRow}:${countCol}${countLastRow}`;
  return {
    stepId: "",
    status: "success",
    message: `Histogram built with ${binEdges.length} bins + overflow. Output: ${outputAddr}.${chartName ? ` Chart: ${chartName}.` : ""}`,
    outputs: { outputRange: outputAddr, ...(chartName ? { chartName } : {}) },
  };
}

/** Letter like "A" → 0; "AA" → 26. */
function columnLetterToIndex(letters: string): number {
  let n = 0;
  for (let i = 0; i < letters.length; i++) n = n * 26 + (letters.charCodeAt(i) - 64);
  return n - 1;
}

/** "A" → "B", "Z" → "AA". */
function nextColumnLetter(letters: string): string {
  const idx = columnLetterToIndex(letters) + 1;
  let n = idx + 1;
  let out = "";
  while (n > 0) {
    const rem = (n - 1) % 26;
    out = String.fromCharCode(65 + rem) + out;
    n = Math.floor((n - 1) / 26);
  }
  return out;
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/histogram.py
"""histogram — FREQUENCY + bin table + optional column chart."""

from __future__ import annotations

import math
import re
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range
from app.execution.utils import parse_number_flexible


def _col_letters_to_index(s: str) -> int:
    n = 0
    for ch in s:
        n = n * 26 + (ord(ch) - ord("A") + 1)
    return n - 1


def _index_to_col_letters(i: int) -> str:
    n = i + 1
    out = ""
    while n > 0:
        rem = (n - 1) % 26
        out = chr(ord("A") + rem) + out
        n = (n - 1) // 26
    return out


def _ensure_2d(v: Any) -> list[list[Any]]:
    if v is None:
        return []
    if not isinstance(v, list):
        return [[v]]
    if not v:
        return []
    if not isinstance(v[0], list):
        return [v] if len(v) > 1 else [[v[0]]]
    return v


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    data_range = params.get("dataRange")
    output_range = params.get("outputRange")
    bins_param = params.get("bins")
    bin_count = params.get("binCount")
    include_chart = bool(params.get("includeChart", True))
    chart_type = params.get("chartType", "columnClustered")
    has_headers = bool(params.get("hasHeaders", True))

    if not data_range or not output_range:
        return {"status": "error", "message": "histogram requires 'dataRange' and 'outputRange'."}

    data_raw = resolve_range(ctx.workbook_handle, data_range)
    data_used = data_raw.current_region if data_raw.count > 1 else data_raw

    raw = _ensure_2d(data_used.value)
    start_row = 1 if has_headers else 0
    nums: list[float] = []
    for r in raw[start_row:]:
        # Tolerate text-stored numbers from CSV imports ($1,234, 50%, etc.)
        n = parse_number_flexible(r[0])
        if n is not None:
            nums.append(n)
    if not nums:
        return {"status": "error", "message": "No numeric values in dataRange."}

    # Bins.
    if bins_param:
        bin_edges = sorted(bins_param)
    else:
        n = len(nums)
        count = bin_count or max(2, math.ceil(math.log2(n) + 1))
        mn, mx = min(nums), max(nums)
        step = (mx - mn) / count if count else 1
        bin_edges = [mn + step * (i + 1) for i in range(count)]

    # Output block header + bin labels + placeholder counts.
    rows = [["Bin", "Count"]]
    for edge in bin_edges:
        rows.append([edge, None])
    rows.append([f">{bin_edges[-1]}", None])

    if ctx.dry_run:
        return {"status": "preview", "message": f"Would build histogram with {len(bin_edges)} bins."}

    out = resolve_range(ctx.workbook_handle, output_range)
    out_top = out[0, 0]
    out_sheet = out.sheet
    out_addr_tail = out.address.split("!")[-1].split(":")[0]
    m = re.match(r"^\$?([A-Z]+)\$?(\d+)$", out_addr_tail)
    if not m:
        return {"status": "error", "message": f"Could not parse output address: {out.address}"}
    out_col_l = m.group(1)
    out_row = int(m.group(2))

    try:
        # Write the block (header + labels + null counts).
        block_target = out_sheet.range(
            (out_row, _col_letters_to_index(out_col_l) + 1),
            (out_row + len(rows) - 1, _col_letters_to_index(out_col_l) + 2),
        )
        block_target.value = rows

        # FREQUENCY array formula into the Count column.
        count_col_l = _index_to_col_letters(_col_letters_to_index(out_col_l) + 1)
        count_first = out_row + 1
        count_last = out_row + len(bin_edges) + 1
        bin_first = out_row + 1
        bin_last = out_row + len(bin_edges)

        # Source data range for FREQUENCY (skip header).
        data_addr = data_used.address
        data_addr_tail = data_addr.split("!")[-1]
        dm = re.match(r"^\$?([A-Z]+)\$?(\d+):\$?([A-Z]+)\$?(\d+)$", data_addr_tail)
        sheet_prefix = data_addr.split("!")[0] + "!" if "!" in data_addr else ""
        if dm:
            c1, r1, c2, r2 = dm.group(1), int(dm.group(2)), dm.group(3), int(dm.group(4))
            data_start = r1 + 1 if has_headers else r1
            data_formula_ref = f"{sheet_prefix}{c1}{data_start}:{c2}{r2}"
        else:
            data_formula_ref = data_addr

        bin_ref = f"{out_col_l}{bin_first}:{out_col_l}{bin_last}"
        formula = f"=FREQUENCY({data_formula_ref},{bin_ref})"

        count_target = out_sheet.range(f"{count_col_l}{count_first}:{count_col_l}{count_last}")
        # Array-formula entry: xlwings `formula_array` for array formulas.
        try:
            count_target.api.FormulaArray = formula
        except Exception:
            count_target.formula = [[formula]] * (len(bin_edges) + 1)
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Write failed: {exc}", "error": str(exc)}

    chart_name: str | None = None
    if include_chart:
        try:
            chart_data = out_sheet.range(f"{out_col_l}{out_row}:{count_col_l}{count_last}")
            chart = out_sheet.charts.add()
            chart.chart_type = "column_clustered" if chart_type == "columnClustered" else "bar_clustered"
            chart.set_source_data(chart_data)
            try:
                chart.api[1].ChartTitle.Text = "Histogram"  # Windows COM path
            except Exception:
                pass
            chart_name = chart.name
        except Exception:
            pass

    output_addr = f"{out_sheet.name}!{out_col_l}{out_row}:{count_col_l}{count_last}"
    return {
        "status": "success",
        "message": f"Histogram built with {len(bin_edges)} bins + overflow. Output: {output_addr}.",
        "outputs": {"outputRange": output_addr, **({"chartName": chart_name} if chart_name else {})},
    }


registry.register("histogram", handler, mutates=True)
